# Assignment 20: Building & Deploying a Content-Based Book Recommendation System

## Name:
**Suryansh Tiwari**

## Objective

The objective of this assignment is to build a **Content-Based Book Recommendation System** using Natural Language Processing techniques. The project focuses on preprocessing textual book information, converting it into numerical vectors using **TF-IDF**, computing similarity using **Cosine Similarity**, and recommending similar books based on user input. Finally, the application will be deployed using **Streamlit** and hosted on **Render**.

---

## Dataset Information

**Dataset Name:** Books Dataset

**Source:** Kaggle

**Dataset Link:**
https://www.kaggle.com/datasets/abdallahwagih/books-dataset

### Features Used

The dataset contains information about thousands of books.

For this project, the following columns will be used:

- Title
- Authors
- Categories
- Description

These columns provide enough textual information to build a content-based recommendation engine.

---

# Part 1 — Data Preprocessing

## Task 1: Load and Understand the Dataset

In this task we will:

- Import required libraries
- Load the dataset using Pandas
- Display the first five records
- Check dataset shape
- Display column names
- Check data types
- Check missing values
- Identify the columns that will be used for recommendations

In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv("data.csv")
df.head()

,isbn13,isbn10,title,subtitle,authors,categories,thumbnail,description,published_year,average_rating,num_pages,ratings_count
0,9780002005883,0002005883,Gilead,NaN,Marilynne Robinson,Fiction,http://books.google.com/books/content?id=KQZCP...,A NOVEL THAT READERS and critics have been eag...,2004.0,3.85,247.0,361.0
1,9780002261982,0002261987,Spider's Web,A Novel,Charles Osborne;Agatha Christie,Detective and mystery stories,http://books.google.com/books/content?id=gA5GP...,A new 'Christie for Christmas' -- a full-lengt...,2000.0,3.83,241.0,5164.0
2,9780006163831,0006163831,The One Tree,NaN,Stephen R. Donaldson,American fiction,http://books.google.com/books/content?id=OmQaw...,Volume Two of Stephen Donaldson's acclaimed se...,1982.0,3.97,479.0,172.0
3,9780006178736,0006178731,Rage of angels,NaN,Sidney Sheldon,Fiction,http://books.google.com/books/content?id=FKo2T...,"A memorable, mesmerizing heroine Jennifer -- b...",1993.0,3.93,512.0,29532.0
4,9780006280897,0006280897,The Four Loves,NaN,Clive Staples Lewis,Christian life,http://books.google.com/books/content?id=XhQ5X...,Lewis' work on the nature of love divides love...,2002.0,4.15,170.0,33684.0


In [3]:
print("Dataset Shape:")
print(df.shape)

print("\nColumn Names:")
print(df.columns)

print("\nDataset Info:")
df.info()

print("\nMissing Values:")
print(df.isnull().sum())

Dataset Shape:
(6810, 12)

Column Names:
Index(['isbn13', 'isbn10', 'title', 'subtitle', 'authors', 'categories',
       'thumbnail', 'description', 'published_year', 'average_rating',
       'num_pages', 'ratings_count'],
      dtype='str')

Dataset Info:
<class 'pandas.DataFrame'>
RangeIndex: 6810 entries, 0 to 6809
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   isbn13          6810 non-null   int64  
 1   isbn10          6810 non-null   str    
 2   title           6810 non-null   str    
 3   subtitle        2381 non-null   str    
 4   authors         6738 non-null   str    
 5   categories      6711 non-null   str    
 6   thumbnail       6481 non-null   str    
 7   description     6548 non-null   str    
 8   published_year  6804 non-null   float64
 9   average_rating  6767 non-null   float64
 10  num_pages       6767 non-null   float64
 11  ratings_count   6767 non-null   float64
dtypes: float64(4

In [4]:
df.sample(5)

,isbn13,isbn10,title,subtitle,authors,categories,thumbnail,description,published_year,average_rating,num_pages,ratings_count
2514,9780399143069,0399143068,Three Complete Novels,NaN,Dick Francis,Fiction,http://books.google.com/books/content?id=2qgAA...,A compilation of three of the author's most re...,1997.0,4.21,707.0,29.0
3428,9780553268942,0553268945,Forging the Darksword,NaN,Margaret Weis;Tracy Hickman,Fiction,http://books.google.com/books/content?id=ttekN...,Born without magical abilities in a magic worl...,1988.0,3.67,391.0,8293.0
6521,9781841957173,1841957178,The Penelopiad,NaN,Margaret Atwood,Fiction,http://books.google.com/books/content?id=HbkV_...,A look at the Odyssey from Penelope's point of...,2005.0,3.69,198.0,26779.0
4041,9780679762706,0679762701,Art & Lies,A Piece for Three Voices and a Bawd,Jeanette Winterson,Fiction,http://books.google.com/books/content?id=0K-vm...,On a high-speed train hurtling through the nea...,1996.0,3.87,240.0,3407.0
6484,9781741042955,174104295X,Ecuador & the Galápagos Islands,NaN,Danny Palmerlee;Michael Grosberg;Carolyn McCarthy,Travel,http://books.google.com/books/content?id=3l4dy...,"Describes the people, customs, geography, and ...",2006.0,3.87,440.0,64.0


In [5]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
isbn13,6810.0,NaN,NaN,NaN,9780676621541.416016,606891056.586608,9780002005883.0,9780330341506.0,9780553108095.5,9780809982553.5,9789042003408.0
isbn10,6810,6810,0002005883,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
title,6810,6398,The Lord of the Rings,11,NaN,NaN,NaN,NaN,NaN,NaN,NaN
subtitle,2381,2009,A Novel,226,NaN,NaN,NaN,NaN,NaN,NaN,NaN
authors,6738,3780,Agatha Christie,37,NaN,NaN,NaN,NaN,NaN,NaN,NaN
categories,6711,567,Fiction,2588,NaN,NaN,NaN,NaN,NaN,NaN,NaN
thumbnail,6481,6481,http://books.google.com/books/content?id=KQZCP...,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
description,6548,6474,No Marketing Blurb,6,NaN,NaN,NaN,NaN,NaN,NaN,NaN
published_year,6804.0,NaN,NaN,NaN,1998.630364,10.484257,1853.0,1996.0,2002.0,2005.0,2019.0
average_rating,6767.0,NaN,NaN,NaN,3.933284,0.331352,0.0,3.77,3.96,4.13,5.0


## Task 2: Text Preprocessing for Recommendation

In this task, the textual information of each book will be cleaned before vectorization.

The following preprocessing steps will be performed:

- Replace missing values with empty strings
- Combine important text features into one column
- Convert text to lowercase
- Remove punctuation and special characters
- Remove stopwords
- Store the processed text in a new column named **clean_text**

In [6]:
import re
import string
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Asus\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [7]:
text_columns=["title","authors","categories","description"]
df[text_columns] = df[text_columns].fillna("")

In [8]:
df["combined_text"]=(df["title"] + " " + df["authors"] + " " + df["categories"] + " " + df["description"])

In [9]:
stop_words=set(stopwords.words("english"))
def clean_text(text):
    text = text.lower()
    text = re.sub(r"[^a-zA-Z\s]", " ", text)
    words = text.split()
    words = [word for word in words if word not in stop_words]
    return " ".join(words)
df["clean_text"] = df["combined_text"].apply(clean_text)

In [10]:
df[["title", "clean_text"]].head()

,title,clean_text
0,Gilead,gilead marilynne robinson fiction novel reader...
1,Spider's Web,spider web charles osborne agatha christie det...
2,The One Tree,one tree stephen r donaldson american fiction ...
3,Rage of angels,rage angels sidney sheldon fiction memorable m...
4,The Four Loves,four loves clive staples lewis christian life ...


In [11]:
print("Missing values in clean_text:")
print(df["clean_text"].isnull().sum())
print("\nEmpty clean_text rows:")
print((df["clean_text"] == "").sum())

Missing values in clean_text:
0

Empty clean_text rows:
0


# Task 3: TF-IDF Vectorization

Machine learning models cannot directly understand textual data. Therefore, the cleaned text must be converted into numerical feature vectors.

In this task, **TF-IDF (Term Frequency–Inverse Document Frequency)** is used to transform the `clean_text` column into a numerical matrix.

TF-IDF assigns higher weights to words that are important in a document while reducing the importance of very common words that appear across many documents.

This TF-IDF matrix will be used in the next task to calculate the similarity between books using **Cosine Similarity**.

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [13]:
tfidf=TfidfVectorizer(max_features=5000,stop_words="english")
tfidf_matrix=tfidf.fit_transform(df["clean_text"])

In [14]:
print("TF-IDf Matrix Shape:", tfidf_matrix.shape)
print("Vocabulary Size:", len(tfidf.get_feature_names_out()))

TF-IDf Matrix Shape: (6810, 5000)
Vocabulary Size: 5000


In [15]:
vocabulary = tfidf.get_feature_names_out()
print("First 30 words in the vocabulary:", vocabulary[:30])

First 30 words in the vocabulary: ['aaron' 'abandoned' 'abbey' 'abby' 'abe' 'abilities' 'ability' 'able'
 'aboard' 'abraham' 'abroad' 'absolutely' 'abstract' 'abuse' 'academy'
 'accept' 'acceptance' 'accepted' 'accepts' 'access' 'accessible'
 'accident' 'accidentally' 'accidents' 'acclaim' 'acclaimed' 'accompanied'
 'accompanies' 'accompany' 'accompanying']


In [16]:
tfidf_df=pd.DataFrame(tfidf_matrix[:5].toarray(), columns=vocabulary)
tfidf_df.iloc[:, :20]

,aaron,abandoned,abbey,abby,abe,abilities,ability,able,aboard,abraham,abroad,absolutely,abstract,abuse,academy,accept,acceptance,accepted,accepts,access
0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.130559,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,0.0
3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.000000,0.0,0.0,0.0


In [17]:
first_book=tfidf_df.iloc[0]
top_words=(first_book[first_book>0].sort_values(ascending=False).head(10))
top_words

son            0.356698
john           0.286567
grandfather    0.256964
critics        0.232078
story          0.186907
readers        0.157317
best           0.145204
recording      0.138686
elder          0.138686
bonds          0.135567
Name: 0, dtype: float64

### Observations

- The TF-IDF vectorizer successfully transformed the cleaned textual data into numerical feature vectors.
- The resulting matrix has **6810 rows** (books) and **5000 columns** (selected vocabulary terms).
- Each book is represented as a 5000-dimensional vector.
- Words with higher TF-IDF scores are considered more representative of a particular book, while very common words receive lower weights.
- These vectors will be used in the next task to compute similarity between books using Cosine Similarity.

# Task 4: Computing Cosine Similarity

After converting the textual information into TF-IDF vectors, the next step is to measure the similarity between books.

Cosine Similarity is one of the most commonly used similarity measures in content-based recommendation systems. It calculates the cosine of the angle between two TF-IDF vectors.

- A similarity score of **1** indicates that two books are identical in terms of their textual representation.
- A similarity score of **0** indicates that the books are completely different.

The resulting similarity matrix will later be used to recommend books that are most similar to a selected book.

In [18]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_sim = cosine_similarity(tfidf_matrix)

In [19]:
print("Cosine Similarity Matrix Shape:", cosine_sim.shape)

Cosine Similarity Matrix Shape: (6810, 6810)


In [20]:
similarity_df=pd.DataFrame(cosine_sim[:5, :5],index=df["title"][:5],columns=df["title"][:5])
similarity_df

title,Gilead,Spider's Web,The One Tree,Rage of angels,The Four Loves
title,,,,,
Gilead,1.000000,0.003468,0.002860,0.006460,0.009151
Spider's Web,0.003468,1.000000,0.020898,0.000000,0.010928
The One Tree,0.002860,0.020898,1.000000,0.004836,0.000000
Rage of angels,0.006460,0.000000,0.004836,1.000000,0.076749
The Four Loves,0.009151,0.010928,0.000000,0.076749,1.000000


In [21]:
print("similarity between Book 1 and Book 2:", cosine_sim[0][1])

similarity between Book 1 and Book 2: 0.0034677531364290686


In [22]:
similar_scores = list(enumerate(cosine_sim[0]))

similar_scores = sorted(
    similar_scores,
    key=lambda x: x[1],
    reverse=True
)

similar_scores[:10]

[(0, np.float64(1.0000000000000004)),
 (918, np.float64(0.21089297235375215)),
 (5295, np.float64(0.2020644304483464)),
 (744, np.float64(0.18373597665350372)),
 (2248, np.float64(0.18212622380617063)),
 (4455, np.float64(0.18119222786475703)),
 (6071, np.float64(0.17852798227155459)),
 (5702, np.float64(0.17174164431491382)),
 (3360, np.float64(0.1645705681171542)),
 (3544, np.float64(0.15361406384496437))]

### Observations

- Cosine Similarity successfully computed the similarity between every pair of books.
- The similarity matrix has dimensions **6810 × 6810**, where each row and column corresponds to a book in the dataset.
- Every diagonal value is equal to **1**, indicating that each book is perfectly similar to itself.
- Higher similarity scores indicate that two books share more similar textual content based on their TF-IDF representations.
- This similarity matrix forms the foundation of the recommendation engine and will be used to retrieve the most relevant books in the next task.

In [23]:
indices = pd.Series(df.index, index=df["title"]).drop_duplicates()
indices.head()

title
Gilead            0
Spider's Web      1
The One Tree      2
Rage of angels    3
The Four Loves    4
dtype: int64

# Task 5: Building the Book Recommendation Function

In this task, a recommendation function is created using the cosine similarity matrix.

The function performs the following steps:

- Accepts a book title as input.
- Finds the corresponding index of the selected book.
- Retrieves the similarity scores of all books.
- Sorts the books based on similarity scores.
- Excludes the selected book itself.
- Returns the top recommended books along with their similarity scores.

In [25]:
def recommend_books(title, top_n=5):
    if title not in indices:
        print("Book not found in the dataset.")
        return

    idx = indices[title]
    similarity_scores = list(enumerate(cosine_sim[idx]))
    similarity_scores = sorted(similarity_scores,key=lambda x: x[1], reverse=True)
    similarity_scores = similarity_scores[1:top_n+1]
    recommendations = []
    for i,score in similarity_scores:
        recommendations.append({
            "Title": df.iloc[i]["title"],
            "Author": df.iloc[i]["authors"],
            "Category": df.iloc[i]["categories"],
            "Similarity Score": round(score, 3)
        })
    return pd.DataFrame(recommendations)

In [30]:
recommend_books("Gilead")

,Title,Author,Category,Similarity Score
0,Go Tell it on the Mountain,James Baldwin,Fiction,0.211
1,Four Baboons Adoring the Sun,John Guare,Drama,0.202
2,The Deep End of the Ocean,Jacquelyn Mitchard,Fiction,0.184
3,Children of the Alley,Najīb Maḥfūẓ,Fiction,0.182
4,The Languages of Pao,Jack Vance,Fiction,0.181


In [27]:
recommend_books("The Hobbit")

Book not found in the dataset.


In [32]:
df["title"].sample(5)

6673                                   White Widow
5239    The Gospel of the Flying Spaghetti Monster
4694                               Sex For Dummies
2800                           Harriet Spies Again
3656                            Persuasive Writing
Name: title, dtype: str

In [34]:
recommend_books("White Widow")

,Title,Author,Category,Similarity Score
0,Works of Jack London,Jack London,Adventure stories,0.217
1,Angry Black White Boy,Adam Mansbach,Fiction,0.213
2,The White House,Debra Hess,History,0.209
3,The Murder of Roger Ackroyd,Agatha Christie,Fiction,0.195
4,One Thousand White Women,Jim Fergus,Fiction,0.187


In [35]:
recommend_books("Persuasive Writing", top_n=5)

,Title,Author,Category,Similarity Score
0,Descriptive Writing,Tara McCarthy,Language Arts & Disciplines,0.676
1,Persuasive Communication,James Brian Stiff;Paul A. Mongeau,Language Arts & Disciplines,0.324
2,Zen in the Art of Writing,Ray Bradbury,Language Arts & Disciplines,0.207
3,Active Literacy Across the Curriculum,Heidi Hayes Jacobs,Education,0.178
4,On Writing,Stephen King,Biography & Autobiography,0.155


# Task 6: Saving Recommendation Artifacts

To use the recommendation system inside a Streamlit application, the processed objects are saved using the `pickle` module.

The following objects are stored:

- Cleaned dataset
- TF-IDF Vectorizer
- Cosine Similarity Matrix

These saved files can later be loaded directly into the Streamlit application without recomputing the preprocessing steps.

In [36]:
import pickle

In [37]:
pickle.dump(df, open("books.pkl","wb"))

In [39]:
pickle.dump(cosine_sim, open("cosine_sim.pkl", "wb"))